File: bench/native-speed/zig/sieve.zig
const std = @import("std");

pub fn main() !void {
    const n: usize = 10_000_000;
    const alloc = std.heap.page_allocator;
    const sieve = try alloc.alloc(i64, n + 1);
    defer alloc.free(sieve);
    var i: usize = 0;
    while (i <= n) : (i += 1) sieve[i] = 1;
    sieve[0] = 0;
    sieve[1] = 0;
    var p: usize = 2;
    while (p * p <= n) : (p += 1) {
        if (sieve[p] == 1) {
            var m: usize = p * p;
            while (m <= n) : (m += p) sieve[m] = 0;
        }
    }
    var count: i64 = 0;
    var k: usize = 0;
    while (k <= n) : (k += 1) count += sieve[k];
    std.debug.print("{d}\n", .{count});
}

File: bench/native-speed/rust/sieve.rs
fn main() {
    let n = 10_000_000usize;
    let mut sieve: Vec<i64> = Vec::new();
    let mut i = 0usize;
    while i <= n { sieve.push(1); i += 1; }
    sieve[0] = 0;
    sieve[1] = 0;
    let mut p = 2usize;
    while p * p <= n {
        if sieve[p] == 1 {
            let mut m = p * p;
            while m <= n { sieve[m] = 0; m += p; }
        }
        p += 1;
    }
    let mut count: i64 = 0;
    let mut k = 0usize;
    while k <= n { count += sieve[k]; k += 1; }
    println!("{}", count);
}
